### PR TITLE
SOR V3 - Use Balancer Maths for V3 Liquidity [fixed]

### LINKED ISSUE #602: SOR V3 - Use Balancer Maths for V3 liquidity
Current pool maths follow Balancer V2 structure.

This task aims to add V3 pool math structure to the SOR, which provides a set of base pool math for add/remove liquidity and requires simply the invariant and "reverse invariant" in order to support new pool types.

It should initially support Weighted and Stable pools, and should be easily extensible to support adding new pool types.

### PR DESCRIPTION
**Pull request by brunoguerios**
Closes #602

### CODE REVIEW

**Comment by brunoguerios** [comment]
There's a final test it should pass before merging, which is to properly find a swap path within `sor-debug.test.ts` going through a stable pool.
It's currently failing due to priceRates being incorrectly scaled.
As soon as this is fixed, tests should pass and then the PR will be good to go.

edit: the priceRate issue is not related to the code itself, but it may be hiding some other issue. That's why I believe it makes sense to wait for that to be fixed before considering this correct/done.

**Comment by brunoguerios** [comment]
Tests are passing ✅ 
The PR should be ready to be merged.
On the next PR I'll try to be a bit more thorough with integration tests in place to give more confidence that the whole thing is working as expected.

File: modules/sor/sorV2/lib/poolsV2/weighted/weightedPoolToken.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+        this.amount = this.amount + amount;
+        this.scale18 = this.amount * this.scalar;
+        return this;
+    }
+
+    public decrease(amount: bigint): TokenAmount {
+        this.amount = this.amount - amount;
+        this.scale18 = this.amount * this.scalar;
+        return this;
+    }
```
Comment (gmbronco) [inline modules/sor/sorV2/lib/poolsV2/weighted/weightedPoolToken.ts:23]: could we use add / sub instead to keep it consistent with number conventions?

File: modules/sor/sorV2/lib/poolsV3/stable/stablePool.test.ts
Comment (gmbronco) [inline modules/sor/sorV2/lib/poolsV3/stable/stablePool.test.ts:1]: could we split the tests into separate parts:

fromPrismaPool - simple unit to make sure we handle edge cases with missing data etc.

swaps - add sor specific factories, so we have expected inputs and don't have to build everything from scratch

are other methods worth unit testing?

**Review by johngrantuk — APPROVED** [review]
Given that this was an update to #603 which had already been given the ok this LGTM.

File: modules/sor/sorV2/lib/poolsV2/weighted/weightedPoolToken.ts (hunk 1 above)
Comment (brunoguerios) [reply modules/sor/sorV2/lib/poolsV2/weighted/weightedPoolToken.ts:23]: I agree it would be a better naming, but since it's the pattern used on all pools, I don't think such a refactor fits in this PR. I created issue #682 to keep track of this.

File: modules/sor/sorV2/lib/poolsV3/stable/stablePool.test.ts
Comment (brunoguerios) [reply modules/sor/sorV2/lib/poolsV3/stable/stablePool.test.ts:1]: I'll split into unit/integration tests over here ✅ 
About creating SOR specific factories so we avoid setting up the whole pool/tokens from scratch on every test, I think it may make sense, but I'd rather wait to see what gets repeated over time and then refactor, instead of trying to anticipate and optimize too soon.
Let me know what you think 😉

Comment (brunoguerios) [reply modules/sor/sorV2/lib/poolsV3/stable/stablePool.test.ts:1]: While splitting the tests, I made them more strict and they highlighted an issue where SOR quote and SDK query are not matching for stable pools. I'll dig deeper to identify what's going on.

Comment (gmbronco) [reply modules/sor/sorV2/lib/poolsV3/stable/stablePool.test.ts:1]: thanks, with the factory it's rather about having separated concerns, than optimisation and i'd like us to keep SOR self-contained as much as possible.

now with the integration test, what's tested there? is it onchain state vs sor math?

File: package.json [hunk 2]
```diff
@@ -22,8 +22,9 @@
         "@aws-sdk/client-cloudwatch": "^3.388.0",
         "@aws-sdk/client-secrets-manager": "^3.195.0",
         "@aws-sdk/client-sqs": "^3.137.0",
+        "@balancer-labs/balancer-maths": "^0.0.6",
         "@balancer-labs/sdk": "github:beethovenxfi/balancer-sdk#beethovenx-master",
-        "@balancer/sdk": "^0.14.0",
+        "@balancer/sdk": "0.20.5",
```
Comment (brunoguerios) [inline package.json:27]: @franzns - can you please confirm this is not an issue for other parts of the code?

File: test/factories/prismaPool.factory.ts [hunk 3]
```diff
@@ -2,78 +2,34 @@ import { Factory } from 'fishery';
...
 import { prismaPoolTokenFactory } from './prismaToken.factory';
 import { createRandomAddress } from '../utils';
+import { Chain, PrismaPoolType } from '@prisma/client';
+import { prismaPoolDynamicDataFactory } from './prismaPoolDynamicData.factory';
 
-export const prismaPoolFactory = Factory.define<PrismaPoolWithDynamic>(({ params }) => {
+class PrismaPoolFactory extends Factory<PrismaPoolWithDynamic> {
+    stable(amp?: string) {
+        return this.params({ type: PrismaPoolType.STABLE, typeData: { amp: amp ?? '10' } });
+    }
```
Comment (brunoguerios) [inline test/factories/prismaPool.factory.ts:11]: @gmbronco - I added a variation of the pool factory so it creates a stable pool with proper parameters. How does that overlap with your new StablePoolFactory?

File: modules/sor/sorV2/lib/poolsV3/stable/stablePool.test.ts
Comment (brunoguerios) [reply modules/sor/sorV2/lib/poolsV3/stable/stablePool.test.ts:1]: Yeap - onchain queries against SOR quotes. They should match exactly. I found and fixed the issue, which was simply some incorrect test pool data (i.e. tokenRates)

File: package.json (hunk 2 above)
Comment (franzns) [reply package.json:27]: the sdk is only used in the SOR part of the API

File: modules/sor/sorV2/lib/poolsV3/stable/stablePool.test.ts
Comment (brunoguerios) [reply modules/sor/sorV2/lib/poolsV3/stable/stablePool.test.ts:1]: btw - all those pool factories are being used on SOR tests only, so I don't think it needs another set of specific factories
let me know if I'm missing anything

File: test/factories/prismaPool.factory.ts (hunk 3 above)
Comment (gmbronco) [reply test/factories/prismaPool.factory.ts:11]: Great, a stable pool factory would work. However, the point is more about design rather than functionality. The `prismaPool` is a database model, and SOR needs its own data format. Ideally, we should separate the SOR test so it doesn't depend on the database and can be worked on independently.

File: modules/sor/sorV2/lib/poolsV3/stable/stablePool.test.ts
Comment (gmbronco) [reply modules/sor/sorV2/lib/poolsV3/stable/stablePool.test.ts:1]: I commented below in the factory thread. Would changing them to SOR-specific ones be a hassle? If so, don’t bother. SOR factories would be very useful for SOR unit testing, and DB factories for integration. This works for now since most tests are integration tests. However, if you want to add more unit tests (which are faster and easier to work with), SOR-specific factories would make more sense.

**Review by gmbronco — APPROVED** [review]

File: modules/sor/sorV2/lib/poolsV3/stable/stablePool.test.ts
Comment (brunoguerios) [reply modules/sor/sorV2/lib/poolsV3/stable/stablePool.test.ts:1]: It would be a bit tricky with the current implementation as the SOR is tightly coupled with DB formats as prismaPools are provided as input. I'll focus on the SOR implementation for now, but I agree that we could benefit from separating concerns in the future.
